Add displaCy data structures to docs

thomashacker opened the pull request:
<!--- Provide a general summary of your changes in the title. -->

## Description
<!--- Use this section to describe your changes. If your changes required
testing, include information about the testing environment and the tests you
ran. If your test fixes a bug reported in an issue, don't forget to include the
issue number. If your PR is still a work in progress, that's totally fine – just
include a note to let us know. -->

This PR adds a new section to the `api/top-level#functions` documentation for displaCy: `Visualizer data structures`. The section shows examples of the three data structures (`Dependency Parsing`, `Entity Recognition`, and `Span Classification`) and their types formatted as tables. To reduce cluttering, the nested dictionary types are formatted as a collapsable table.

The idea for this PR initially came from this discussion https://github.com/explosion/spaCy/pull/10950
The types are based on [this gist](https://gist.github.com/pmbaumgartner/85b3a0256131e86fbdd78a8b4060d361) by @pmbaumgartner  

### Types of change
<!-- What type of change does your PR cover? Is it a bug fix, an enhancement
or new feature, or a change to the documentation? -->
Documentation

## Checklist
<!--- Before you submit the PR, go over this checklist and make sure you can
tick off all the boxes. [] -> [x] -->
- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

pmbaumgartner commented on website/docs/api/top-level.mdx:
```diff
@@ -340,6 +340,128 @@ use with the `manual=True` argument in `displacy.render`.
+You can also use displaCy's data format to manually render data. This can be
+useful if you want to visualize output from other libaries. You can find
+examples of displaCy's data format on the
+[usage page](/usage/visualizers#manual-usage).
```
Do you think we should still link to the Usage page when the same examples are now in the sidebar? I think it might be more confusing to a user that they would be reading and click this link to be brought away from the documentation.

thomashacker replied:
I agree! I'll remove the link

thomashacker replied:
Removed the link 😇

pmbaumgartner commented on website/docs/api/top-level.mdx:
```diff
@@ -340,6 +340,128 @@ use with the `manual=True` argument in `displacy.render`.
+
+| Dictionary Key | Description                |
+| -------------- | -------------------------- |
+| `start`        | Start index. ~~int~~       |
```
This is more of a general comment we should check across all keys and descriptions: I think the descriptions should match their equivalent data structure description in the spaCy API. For example, in the `Span` [docs](https://spacy.io/api/span) we describe the `start` attribute as "The index of the first token of the span."

thomashacker replied:
Ah, I didn't realize! Thanks, I agree!

thomashacker replied:
I tried to stay consistent with the code base now, but at some attributes, I'm still a bit unsure.

pmbaumgartner commented on website/docs/api/top-level.mdx:
```diff
@@ -340,6 +340,128 @@ use with the `manual=True` argument in `displacy.render`.
+| `start_token`  | Start token. ~~int~~       |
+| `end_token`    | End token. ~~int~~         |
+| `label`        | Label of the span. ~~str~~ |
+| `kb_id`        | Knowledgebase ID. ~~str~~  |
```
```suggestion
| `kb_id`        | `KnowledgeBase` ID. ~~str~~  |
```

Should we style this like this and maybe link to the [KnowledgeBase](https://spacy.io/api/kb) docs as well?

thomashacker replied:
Ah, yes! I agree

thomashacker replied:
Done!

pmbaumgartner reviewed (changes requested):
My in-line code comments are a little out of order so let me summarize them in a comment:

- I don't think we should link back to the Usage docs since we provide those examples in the sidebar
- Can we find a better way to make it clear the accordions have information about the sub-structures in each data structure?
- We should be consistent with the descriptions for the values/attributes where they exist in the spaCy API.

thomashacker commented on website/docs/api/top-level.mdx:
```diff
@@ -340,6 +340,128 @@ use with the `manual=True` argument in `displacy.render`.
+| `arcs`         | List of arcs. ~~List[Dict[str, Any]]~~    |
+| `settings`     | Visualization options. ~~Dict[str, Any]~~ |
+
+<Accordion title="Word data structure">
```
I agree, it might need more clarification and the naming needs to be clearer! I'll change it

thomashacker replied:
I've added `see structure below` . I think it makes it more clearer now

pmbaumgartner commented on website/docs/api/top-level.mdx:
```diff
@@ -340,6 +340,127 @@ use with the `manual=True` argument in `displacy.render`.
+| `start`        | The index of the first token of the entity. ~~int~~ |
+| `end`          | The index of the last token of the entity. ~~int~~  |
+| `label`        | Label attached to the entity. ~~str~~               |
+| `kb_id`        | `KnowledgeBase` ID. ~~str~~                         |
+| `kb_url`       | `KnowledgeBase` URL. ~~str~~                        |
```
I think these are both not required, so we might want to type as `Optional` to be consistent. However, this isn't exactly right because we're using `Optional` here to represent that the key doesn't need to be present, not that the type of the value is `Union[str, None]`.

Maybe we want another way to indicate the key is optional rather than the type. Like placing `(optional)` after the key name in the `Dictionary Key` column?

This is one area it would be good to have feedback on. I'm not sure if there's another place in the docs we have the same issue and what the practice is.  cc @svlandeg

svlandeg replied:
This is basically similar to kwd-args that are optional but not necessarily of type `Optional`. We denote those with a specific divider, cf https://spacy.io/api/language#init for example - so perhaps do something similar to that?

thomashacker replied:
This is great! Thanks, I'll try to add them with a divider

thomashacker replied:
I've added the `Optional` divider for the `KnowledgeBase` parameters and also for the `setting` parameter.

thomashacker commented:
@pmbaumgartner What was the reason for removing the docs label from this PR? 😢

pmbaumgartner commented:
> @pmbaumgartner What was the reason for removing the `docs` label from this PR? 😢

I must have misclicked or used a keyboard shortcut accidentally. Sorry!

svlandeg commented:
@pmbaumgartner: is your "changes requested" review still valid? (as discussed internally we don't usually use this feature as typically we'd go by the open review comments to determine whether something requires changes or not, but because you used it earlier I want to double check with you and not just dismiss the review as that looks weird)

pmbaumgartner commented:
> @pmbaumgartner: is your "changes requested" review still valid? (as discussed internally we don't usually use this feature as typically we'd go by the open review comments to determine whether something requires changes or not, but because you used it earlier I want to double check with you and not just dismiss the review)

I wasn't aware this was still blocking. I dismissed the review, but haven't had time to review the recent changes yet, so I'll look again today.

pmbaumgartner commented:
In the `ENT example data structure`, `title` has a value of `None`. Shouldn't its actual type be `Optional[str]` then then? Same for Spans?

And it looks like `title` isn't a field on the `deps` data, only `ents` and `spans`, is that right? I think we might want to open an issue for that, if that's the case.

thomashacker commented:
- Adjusted the type of the `title` key in the docs
- Added tests that check `displaCy.render` with manual data
- Implemented `title` key to dependency visualizations and adjusted docs

pmbaumgartner commented:
👌 love the tests. I don't have feedback on anything else at this time.

svlandeg commented:
Closing in favour of #12875 (PR redone to target `master`)
